### PR TITLE
Fix for: Colordialog is incorrectly positioned when used with another dialog

### LINKED ISSUE #3559: Colordialog is incorrectly positioned when used with another dialog
## Type of report

Bug

## Provide detailed reproduction steps (if any)

1. Open CKEditor4 sample.
2. Open color dialog using colorbutton plugin and change its position.
3. Close color dialog.
4. Add a new table to the editor.
5. Open color dialog again from cell properties dialog.

### Expected result

The color dialog is responsive, correctly positioned and overflow table properties dialog. See 4.12.1 how it worked.

### Actual result

The color dialog is hidden under the table properties dialog. The table dialog is not dimmed. The color dialog is not responsive at all.

colorbutton

## Other details

* CKEditor version: regression provided with 4.13.0

### PR DESCRIPTION
**Pull request by Dumluregn**
## What is the purpose of this pull request?

Bug fix

## Does your PR contain necessary tests?

All patches which change the editor code must include tests. You can always read more
on [PR testing](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_contributing_code.html#tests),
[how to set the testing environment](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html) and
[how to create tests](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html#creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [x] Unit tests
- [x] Manual tests

## What is the proposed changelog entry for this pull request?

```
[#3559](https://github.com/ckeditor/ckeditor4/issues/3559): Fix for: [Color Dialog](https://ckeditor.com/cke4/addon/colordialog) is incorrectly positioned when used with another dialog.
```

## What changes did you make?

After the fix cell properties dialog is properly fading when color dialog is opened.

Closes #3559.

### CODE REVIEW

File: tests/plugins/tabletools/manual/colordialogposition.html [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+			<tr>
+				<td>2.1</td>
+				<td>2.2</td>
+			</tr>
+		</tbody>
+	</table>
+</div>
+
+<script>
+	if ( bender.tools.env.mobile ) {
```
Comment (jacekbogdanski) [inline tests/plugins/tabletools/manual/colordialogposition.html]: Would be nice to also add manual test for mobiles, as the regression which has been introduced here is caused by changes dedicated to mobiles. I think that if we change test steps with dev console, it should be possible to do it in the same test file, WDYT?

File: plugins/tabletools/plugin.js [hunk 2]
```diff
@@ -531,37 +552,41 @@
...
+			commonAncestor;
 
 		// Invalid merge request if:
 		// 1. In batch mode despite that less than two selected.
 		// 2. In solo mode while not exactly only one selected.
 		// 3. Cells distributed in different table groups (e.g. from both thead and tbody).
-		var commonAncestor;
-		if ( ( mergeDirection ? cells.length != 1 : cells.length < 2 ) || ( commonAncestor = selection.getCommonAncestor() ) && commonAncestor.type == CKEDITOR.NODE_ELEMENT && commonAncestor.is( 'table' ) )
+		if ( ( mergeDirection ? cells.length != 1 : cells.length < 2 ) ||
+			( commonAncestor = selection.getCommonAncestor() ) &&
```
[the remark this answers is not in the document]
Comment (Dumluregn) [reply plugins/tabletools/plugin.js]: I looked at this conditional statement a bit closer and it seems a bit weird... Normally I'd say that extracting this may be risky, because if `( mergeDirection ? cells.length != 1 : cells.length < 2 )` is true, the part after `||` (so with `commonAncestor`) will not be calculated and maybe it shouldn't because of some possible bugs. But... later we have `commonAncestor.type == CKEDITOR.NODE_ELEMENT` and `commonAncestor.is( 'table' )` that should throw `TypeError` if `( commonAncestor = selection.getCommonAncestor() )` didn't run, so it seems that the first condition is always false 🤔 So after all I think your suggestion may actually save us some super-duper-edge-case error in the future 😂 (of course please correct me if I'm wrong).

File: tests/plugins/tabletools/manual/colordialogposition.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+  ### Unexpected:
+  Any dialog doesn't disappear and makes editor unavailable.
+
+1. Using console check also independent dialogs ordering, e.g. type:
+```
+CKEDITOR.instances.editor.execCommand( 'link' );
+CKEDITOR.instances.editor.execCommand( 'colordialog' );
+CKEDITOR.instances.editor.execCommand( 'link' );
+```
```
[the remark this answers is not in the document]
Comment (Dumluregn) [reply tests/plugins/tabletools/manual/colordialogposition.md]: Thanks for the tip, I will do it 👍 But I think we can only use one button opening all dialogs (because of dialog cover).

File: tests/plugins/tabletools/manual/colordialogposition.html (hunk 1 above)
Comment (Dumluregn) [reply tests/plugins/tabletools/manual/colordialogposition.html]: Yup, should work 👍

File: plugins/tabletools/plugin.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (jacekbogdanski) [reply plugins/tabletools/plugin.js]: So, we already see  that this code is unreadable as hell :smile: Look how it splits:

```javascript
var isBatchMode = mergeDirection ? cells.length != 1 : cells.length < 2;

if ( isBatchMode ) {
  return false;
}

var commonAncestor = selection.getCommonAncestor();

if ( isTableElement( commonAncestor ) ) {
  return false;
}

function isTableElement( node ) {
  return node && node.type === CKEDITOR.NODE_ELEMENT && node.is( 'table' );
}
```

It's the same code as this crazy `if` - although please, make sure to test it and verify if it's correct, I could also make some mistake.

[the remark this answers is not in the document]
Comment (jacekbogdanski) [reply plugins/tabletools/plugin.js]: About the case with:
> But... later we have commonAncestor.type == CKEDITOR.NODE_ELEMENT and commonAncestor.is( 'table' ) that should throw TypeError if ( commonAncestor = selection.getCommonAncestor() ) didn't run, so it seems that the first condition is always false thinking
look on brackets here: `( commonAncestor = selection.getCommonAncestor() ) && ...`, code after && won't be evaulated if `commonAncestor` result with `null`.

File: plugins/tabletools/dialogs/tableCell.js [hunk 4]
```diff
@@ -538,4 +513,31 @@ CKEDITOR.dialog.add( 'cellProperties', function( editor ) {
...
+			selectedCell.removeStyle( property );
+		}
+
+		// As this is used only in these two cases now, we can name them explicitly for clarity.
+		// If we ever need to generalise it, CKEDITOR.tools.cssStyleToDomStyle() may come in handy.
+		if ( property == 'background-color' ) {
+			selectedCell.removeAttribute( 'bgColor' );
+		} else if ( property == 'border-color' ) {
+			selectedCell.removeAttribute( 'borderColor' );
+		}
```
[the remark this answers is not in the document]
Comment (Dumluregn) [reply plugins/tabletools/dialogs/tableCell.js]: I'm afraid not, as `CKEDITOR.tools.cssStyleToDomStyle( 'background-color' )` returns `backgroundColor` instead of `bgColor` 😕

File: plugins/tabletools/plugin.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (Dumluregn) [reply plugins/tabletools/plugin.js]: I think the first early return is unsafe - we can simplify the original `if` (overindentation is deliberate 😛): 
```javascript
if ( 
    ( mergeDirection ? cells.length != 1 : cells.length < 2 ) || 
    ( commonAncestor = selection.getCommonAncestor() ) && 
    commonAncestor.type == CKEDITOR.NODE_ELEMENT && 
    commonAncestor.is( 'table' ) 
    ) {
    return false;
}
```
to:
```javascript
if ( a || b && c && d )
```
So even if `a == true` we still need `c == true` and `d == true`.
But... `c` and `d` can only be `true` if `b == true` (or we get TypeError as I mentioned before). In fact, `a == false` doesn't change anything if all `b, c, d` are equal to `true` 😱So it seems to me that all that code can be replaced with just:

```javascript
var commonAncestor = selection.getCommonAncestor(),
    ancestorIsNode = commonAncestor && ( commonAncestor.type == 
    CKEDITOR.NODE_ELEMENT ),
    ancestorIsTable = ancestorIsNode && commonAncestor.is( 'table' );

if ( ancestorIsTable ) {
    return;
}
```

[the remark this answers is not in the document]
Comment (jacekbogdanski) [reply plugins/tabletools/plugin.js]: Let's slow down by removing the first condition :smile: 

```javascript
if ( 
    ( mergeDirection ? cells.length != 1 : cells.length < 2 ) ||  // If true, the rest of the code won't evaluate and function returns false. If false, we are going deeper.
    ( commonAncestor = selection.getCommonAncestor() ) && // If null, the rest of the code won't evaluate. If node, going deeper.
    commonAncestor.type == CKEDITOR.NODE_ELEMENT &&  // If false, nothing. True - deeper.
    commonAncestor.is( 'table' ) // False, nothing. True - function returns false. 
    ) {
    return false;
}
```

However, if you are still not sure about this refactoring, let's leave it as, so we won't accidentally introduce any bug.

[the remark this answers is not in the document]
Comment (Dumluregn) [reply plugins/tabletools/plugin.js]: Yeah, you are totally right, I misanalysed this... Still as this piece of code isn't covered by tests (removing any condition or even entire `if` statement doesn't cause any test to fail) it is probably best to leave it as is to merge this PR as quickly as possible.

File: plugins/tabletools/dialogs/tableCell.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (jacekbogdanski) [reply plugins/tabletools/dialogs/tableCell.js]: That's a pity. I will remove additional command in that case, as it doesn't seem correct anymore.

**Comment by Dumluregn** [comment]
I can fix it but I wonder if it's really an issue - more important than dialogs order is the fact that the one on top is accessible, when it's closed the second one becomes active etc. So maybe instead of messing around with commands we could change **Expected** to all dialogs work properly instead, WDYT?

**Comment by jacekbogdanski** [comment]
I have a feeling that order is important here, note that you actually update code around logic responsible for correct dialog order, which may be messed up if `z-index` is incorrectly updated (on the wrong element). I'm wondering if it won't be simpler to update the test with 3 buttons (all different dialogs) with some crazy high z-index so they are above dialog. Not sure how it will look/being accessible.

**Comment by Dumluregn** [comment]
The bug you reported [here](https://github.com/ckeditor/ckeditor4/pull/3572#discussion_r343065713) has occured before, so I reported it in #3638 to handle it separately if we decide it's important. For now when you click the same button in test more than once it will fail the test, so I included a note to click each one only once.
